src/AvaloniaEdit.TextMate/TextEditorModel.cs

```diff
@@ -124,22 +128,46 @@ private void DocumentOnChanged(object sender, DocumentChangeEventArgs e)
 
                 if (startLine == 0)
                 {
-                    InvalidateLineRange(startLine, endLine);
+                    SetInvalidRange(startLine, endLine);
                     return;
                 }
 
                 // some grammars (JSON, csharp, ...)
                 // need to invalidate the previous line too
 
-                InvalidateLineRange(startLine - 1, endLine);
+                SetInvalidRange(startLine - 1, endLine);
             }
             catch (Exception ex)
             {
                 _exceptionHandler?.Invoke(ex);
             }
         }
 
-        void TokenizeViewPort()
+        private void SetInvalidRange(int startLine, int endLine)
+        {
+            if (!_document.IsInUpdate)
+            {
+                InvalidateLineRange(startLine, endLine);
+                return;
+            }
+
+            // we're in a document change, store the max invalid range
+            if (_invalidRange == null)
+                _invalidRange = new InvalidLineRange(startLine, endLine);
+
+            _invalidRange.SetInvalidRange(startLine, endLine);
+        }
+
+        void DocumentOnUpdateFinished(object sender, EventArgs e)
+        {
+            if (_invalidRange == null)
+                return;
+
+            InvalidateLineRange(_invalidRange.StartLine, _invalidRange.EndLine);
+            _invalidRange = null;
```

Comment:
If `InvalidateLineRange` fails `_invalidRange` will never be set to null. Use try/finally:
```suggestion
            try
            {
                 InvalidateLineRange(_invalidRange.StartLine, _invalidRange.EndLine);
            }
            {
            finally          
            {           
                _invalidRange = null;
            }  
```